[SPARK-17993][SQL] Fix Parquet log output redirection

mallman opened the pull request:
(Link to Jira issue: https://issues.apache.org/jira/browse/SPARK-17993)
## What changes were proposed in this pull request?

PR #14690 broke parquet log output redirection for converted partitioned Hive tables. For example, when querying parquet files written by Parquet-mr 1.6.0 Spark prints a torrent of (harmless) warning messages from the Parquet reader:

```
Oct 18, 2016 7:42:18 PM WARNING: org.apache.parquet.CorruptStatistics: Ignoring statistics because created_by could not be parsed (see PARQUET-251): parquet-mr version 1.6.0
org.apache.parquet.VersionParser$VersionParseException: Could not parse created_by: parquet-mr version 1.6.0 using format: (.+) version ((.*) )?\(build ?(.*)\)
    at org.apache.parquet.VersionParser.parse(VersionParser.java:112)
    at org.apache.parquet.CorruptStatistics.shouldIgnoreStatistics(CorruptStatistics.java:60)
    at org.apache.parquet.format.converter.ParquetMetadataConverter.fromParquetStatistics(ParquetMetadataConverter.java:263)
    at org.apache.parquet.hadoop.ParquetFileReader$Chunk.readAllPages(ParquetFileReader.java:583)
    at org.apache.parquet.hadoop.ParquetFileReader.readNextRowGroup(ParquetFileReader.java:513)
    at org.apache.spark.sql.execution.datasources.parquet.VectorizedParquetRecordReader.checkEndOfRowGroup(VectorizedParquetRecordReader.java:270)
    at org.apache.spark.sql.execution.datasources.parquet.VectorizedParquetRecordReader.nextBatch(VectorizedParquetRecordReader.java:225)
    at org.apache.spark.sql.execution.datasources.parquet.VectorizedParquetRecordReader.nextKeyValue(VectorizedParquetRecordReader.java:137)
    at org.apache.spark.sql.execution.datasources.RecordReaderIterator.hasNext(RecordReaderIterator.scala:39)
    at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:102)
    at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:162)
    at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:102)
    at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.scan_nextBatch$(Unknown Source)
    at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
    at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
    at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:372)
    at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:231)
    at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:225)
    at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:803)
    at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:803)
    at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
    at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:319)
    at org.apache.spark.rdd.RDD.iterator(RDD.scala:283)
    at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
    at org.apache.spark.scheduler.Task.run(Task.scala:99)
    at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:282)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
```

This only happens during execution, not planning, and it doesn't matter what log level the `SparkContext` is set to. That's because Parquet (versions < 1.9) doesn't use slf4j for logging. Note, you can tell that log redirection is not working here because the log message format does not conform to the default Spark log message format.

This is a regression I noted as something we needed to fix as a follow up.

It appears that the problem arose because we removed the call to `inferSchema` during Hive table conversion. That call is what triggered the output redirection.

## How was this patch tested?

I tested this manually in four ways:
1. Executing `spark.sqlContext.range(10).selectExpr("id as a").write.mode("overwrite").parquet("test")`.
2. Executing `spark.read.format("parquet").load(legacyParquetFile).show` for a Parquet file `legacyParquetFile` written using Parquet-mr 1.6.0.
3. Executing `select * from legacy_parquet_table limit 1` for some unpartitioned Parquet-based Hive table written using Parquet-mr 1.6.0.
4. Executing `select * from legacy_partitioned_parquet_table where partcol=x limit 1` for some partitioned Parquet-based Hive table written using Parquet-mr 1.6.0.

I ran each test with a new instance of `spark-shell` or `spark-sql`.

Incidentally, I found that test case 3 was not a regression—redirection was not occurring in the master codebase prior to #14690.

I spent some time working on a unit test, but based on my experience working on this ticket I feel that automated testing here is far from feasible.

cc @ericl @dongjoon-hyun

Commit:
[SPARK-17993][SQL] Perform Parquet log output redirection when
constructing an instance of `ParquetFileFormat`. Before, it was occurring
as part of the call to `inferSchema` and in `prepareWrite`, however not
all Parquet access occurs through one of those methods. We add this
redirection to the constructor to ensure that any instantiation of
`ParquetFileFormat` triggers log redirection if it hasn't already
happened

SparkQA commented:
**[Test build #67150 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67150/consoleFull)** for PR 15538 at commit [`6101b83`](https://github.com/apache/spark/commit/6101b837eec731ad83904593da123c4b30a25f15).

SparkQA commented:
**[Test build #67150 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67150/consoleFull)** for PR 15538 at commit [`6101b83`](https://github.com/apache/spark/commit/6101b837eec731ad83904593da123c4b30a25f15).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67150/
Test FAILed.

ericl commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala:
```diff
@@ -56,6 +56,7 @@ class ParquetFileFormat
   with DataSourceRegister
   with Logging
   with Serializable {
+  ParquetFileFormat.redirectParquetLogs
```
Should we only invoke this once per jvm?

mallman replied:
Essentially, yes.

mallman replied:
The tricky bits are _when_, _where_ and _how_.

ericl replied:
If you moved this into an object that was referenced by this class on init, would that suffice?

mallman replied:
I'm not sure I understand what you're suggesting. Can you elaborate further?

srowen replied:
You should just be able to put this in the class init for `ParquetFileFormat`, by adding this to the body of `object ParquetFileFormat` below.

mallman replied:
That doesn't work. To wit: http://stackoverflow.com/questions/2347107/what-is-scala-equivalent-of-javas-static-block.

srowen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala:
```diff
@@ -718,7 +717,8 @@ object ParquetFileFormat extends Logging {
+   * `redirectParquetLogsViaSLF4J`.
    */
-  def redirectParquetLogs(): Unit = {}
+  private def redirectParquetLogs(): Unit = {}
```
Probably a dumb question but this is a no-op? how does this work?

mallman replied:
If I just call `redirectParquetLogsViaSLF4J` I get a compilation error:

```
[error] [warn] /Volumes/VideoAmpCS/msa/workspace/spark-master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala:59: a pure expression does nothing in statement position; you may be omitting necessary parentheses
[error] [warn]   ParquetFileFormat.redirectParquetLogsViaSLF4J
```

srowen replied:
Yes, because it's not a method. I'm wondering what this method does, and I think it exists only to cause the `object` to initialize. That's awkward, but it already exists I guess. Is there really no way to implement a static init block in Scala? hm. If so then I support this change. It's actually being moved to a spot where it will be called earlier, and, be called fewer times, right?

ericl replied:
Maybe it would be clearer to call it once with an atomic boolean guard:

```
if (ParquetFileFormat.logsRedirected.compareAndSwap(false, true)) {
  ParquetFileFormat.redirectParquetLogsViaSLF4J()
}
```

Then you could test that the flag is set after constructing a ParquetFileFormat. That's obviously not a great test but probably good enough.

mallman replied:
I'll try to clean things up in another commit. If nothing else, it's clear I need to add some code comments.

mallman commented:
I could use some advice on writing a unit test for this. Do you guys know if there is a precedent in the codebase that covers a situation like this? I'd like to reuse existing code if possible.

Commit:
Improve code documentation to clarify the reasoning behind this strategy

mallman commented:
I pushed a commit to improve the documentation. I also removed a couple of unused imports (boy scout rule).

SparkQA commented:
**[Test build #67279 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67279/consoleFull)** for PR 15538 at commit [`099afca`](https://github.com/apache/spark/commit/099afca41893598e2914c161ae1a77b7277f0df4).

SparkQA commented:
**[Test build #67279 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67279/consoleFull)** for PR 15538 at commit [`099afca`](https://github.com/apache/spark/commit/099afca41893598e2914c161ae1a77b7277f0df4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67279/
Test PASSed.

srowen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala:
```diff
@@ -718,7 +722,9 @@ object ParquetFileFormat extends Logging {
+   * only require that `redirectParquetLogsViaSLF4J` is initialized. This accomplishes that task.
    */
-  def redirectParquetLogs(): Unit = {}
+  private def ensureParquetLogRedirection(): Unit = {}
```
OK this is better, though on a second look, it seems less indirect to either:

a) use a `lazy val` above and simply do anything with that reference that compiles, like log it or something. It didn't before because just referencing a field and doing nothing else will cause a warning, or

b) make it a method that executes once with a simple `AtomicBoolean` guard or boolean + synchronization

I suppose I'm wondering whether even this mechanism is guaranteed to work.

srowen replied:
I suppose it's no worse than the existing setup, though surely there's a better way to get once-per-class execution. Would you try one more thing: 

`private def ensureParquetLogRedirection(): Unit = redirectParquetLogsViaSLF4J`

_That_ would at least connect the things that are supposed to be connected though it's all a no-op and would look less wonky to me.

If that still generates a warning OK let's merge this.

mallman replied:
I believe the log output redirection is working. The mystery here is why the test log level threshold is not having an effect in Jenkins.

Nevertheless I'll make the change and push it.

mallman commented:
I've spent a couple hours today working on a unit test which captures stdout during a parquet write operation to validate that it has no parquet logging output. I haven't got it working yet, but I'll spend some more time on it tomorrow.

ericl commented:
also LGTM as-is; it's polluting test output which is kind of annoying. Testing the log output might be a little overkill since this is not critical functionality and has an upstream fix.

mallman commented:
@ericl What do you mean it's polluting test output?

ericl commented:
I see ~10k lines of "CorruptStatistics" stack traces in the jenkins log. Though, it also seems to show up in the test logs for this pr, so maybe it wouldn't affect this.

mallman commented:
I'll take a closer look at that.

Commit:
Set parquet library logging threshold to ERROR for testing

SparkQA commented:
**[Test build #67542 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67542/consoleFull)** for PR 15538 at commit [`dec65c7`](https://github.com/apache/spark/commit/dec65c72e43bd3bcad90dc6f8d79652b2f3c5463).

mallman commented:
The "CorruptStatistics" stack traces (which I agree are really annoying) are being logged because parquet logs them at the WARN level, and Spark's default logging threshold when running tests is WARN.

I've pushed a commit to set the parquet log threshold to ERROR for testing. This is actually the default configuration for a Spark deployment: https://github.com/apache/spark/blob/v2.0.1/conf/log4j.properties.template#L35.

SparkQA commented:
**[Test build #67542 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67542/consoleFull)** for PR 15538 at commit [`dec65c7`](https://github.com/apache/spark/commit/dec65c72e43bd3bcad90dc6f8d79652b2f3c5463).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67542/
Test FAILed.

mallman commented:
I like this test failure:

```
org.apache.spark.sql.sources.CreateTableAsSelectSuite.(It is not a test)
```

Anyway, I don't think this is related to this PR.

mallman commented:
So raising the log threshold looks like it didn't do anything for Jenkins, but when I run the tests locally it does just the trick. \*sigh\*

Anyway, might as well push a rebase and see what happens.

SparkQA commented:
**[Test build #67552 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67552/consoleFull)** for PR 15538 at commit [`02df8c2`](https://github.com/apache/spark/commit/02df8c273ffac794bfb5bff6f3cc0ab9532264f9).

SparkQA commented:
**[Test build #67552 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67552/consoleFull)** for PR 15538 at commit [`02df8c2`](https://github.com/apache/spark/commit/02df8c273ffac794bfb5bff6f3cc0ab9532264f9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67552/
Test PASSed.

mallman commented:
I'm still seeing the torrent of `CorruptStatistics` errors in the Jenkins build log, even though I don't see them running the tests locally with sbt. Maybe it's a maven versus sbt build issue.

Anyway, having spent a few hours trying to get a reliable, effective unit test for this functionality, I don't think it's going to happen.

Can you guys let me know if you have a blocking concern around this PR? Otherwise, can we please merge this to master? I'm reluctant to spend more time on this.

Cheers.

Commit:
Set def ensureParquetLogRedirection(): Unit =
redirectParquetLogsViaSLF4J

SparkQA commented:
**[Test build #67658 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67658/consoleFull)** for PR 15538 at commit [`1fc3c93`](https://github.com/apache/spark/commit/1fc3c939f0e1611cb7f311a33083b234526409ed).

SparkQA commented:
**[Test build #67658 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67658/consoleFull)** for PR 15538 at commit [`1fc3c93`](https://github.com/apache/spark/commit/1fc3c939f0e1611cb7f311a33083b234526409ed).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67658/
Test FAILed.

mallman commented:
Looks like the test failed for reasons unrelated to this PR. Can someone trigger a retest, please?

ericl commented:
Jenkins retest this please

On Thu, Oct 27, 2016, 12:34 PM Michael Allman <notifications@github.com>
wrote:

Looks like the test failed for reasons unrelated to this PR. Can someone
trigger a retest, please?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/apache/spark/pull/15538#issuecomment-256745918>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAA6SrEooeo7aeoob_0UR5YycU3QSCVDks5q4PyqgaJpZM4KaZR_>
.

srowen commented:
Jenkins add to whitelist

SparkQA commented:
**[Test build #3377 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3377/consoleFull)** for PR 15538 at commit [`1fc3c93`](https://github.com/apache/spark/commit/1fc3c939f0e1611cb7f311a33083b234526409ed).

SparkQA commented:
**[Test build #67707 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67707/consoleFull)** for PR 15538 at commit [`1fc3c93`](https://github.com/apache/spark/commit/1fc3c939f0e1611cb7f311a33083b234526409ed).

SparkQA commented:
**[Test build #3377 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3377/consoleFull)** for PR 15538 at commit [`1fc3c93`](https://github.com/apache/spark/commit/1fc3c939f0e1611cb7f311a33083b234526409ed).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #67707 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67707/consoleFull)** for PR 15538 at commit [`1fc3c93`](https://github.com/apache/spark/commit/1fc3c939f0e1611cb7f311a33083b234526409ed).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67707/
Test PASSed.

srowen reviewed (approved):
Great, if this still works, this looks more understandable to me.

mallman commented:
@srowen I ran my manual tests for this build and they worked as expected. Can you merge this PR?

mallman commented:
Actually, this may not be working for remote executors. I tested this patch running in local mode, but in running a version of this with actual remote executors I'm seeing the original parquet log output.

Let me take a closer look before merging this PR.

srowen commented:
OK, well whatever works here, even if somehow the previous version is what's needed to get it to work.

Commit:
Call ParquetFileFormat.ensureParquetLogRedirection when deserializing
ParquetFileFormat

Commit:
Call ParquetFileFormat.ensureParquetLogRedirection when deserializing
Parquet output writer factories

SparkQA commented:
**[Test build #67794 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67794/consoleFull)** for PR 15538 at commit [`e544397`](https://github.com/apache/spark/commit/e544397290911d0d8b2e58992aa1929f8c182784).

SparkQA commented:
**[Test build #67795 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67795/consoleFull)** for PR 15538 at commit [`7f858c6`](https://github.com/apache/spark/commit/7f858c69fa6911af8d57295d485c17c0097a005a).

mallman commented:
Getting this redirection to work on remote executors was quite involved. The additional complexities derive from the following:

1. Java doesn't call the default (or any) constructor in deserialization.
2. Spark doesn't use `ParquetFileFormat` at all in the write path.

To address the first issue, I added a deserialization hook to `ParquetFileFormat` which calls `ParquetFileFormat.ensureParquetLogRedirection`. To address the second, I added deserialization hooks to a couple of classes the driver serializes when it sends the output job to the executors.

This is messy, but it's the simplest, most direct way I found to tackle these complexities.

Hopefully all of this awfulness can be removed when Spark updates its dependency on parquet-mr to the 1.9 release, which uses slf4j for logging.

SparkQA commented:
**[Test build #67794 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67794/consoleFull)** for PR 15538 at commit [`e544397`](https://github.com/apache/spark/commit/e544397290911d0d8b2e58992aa1929f8c182784).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67794/
Test PASSed.

SparkQA commented:
**[Test build #67795 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67795/consoleFull)** for PR 15538 at commit [`7f858c6`](https://github.com/apache/spark/commit/7f858c69fa6911af8d57295d485c17c0097a005a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67795/
Test PASSed.

srowen commented:
If this is necessary, then isn't it simpler to leave the log configuration call where it was, so that it doesn't depend on the constructor? that wasn't the actual problem was it, just the logging? if the problem is that it fires repeatedly, then a flag could be used to record that it has executed?

ericl commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala:
```diff
@@ -55,6 +56,21 @@ class ParquetFileFormat
+  // method.
+  private def readObject(in: ObjectInputStream): Unit = {
+    in.defaultReadObject
+    ParquetFileFormat.ensureParquetLogRedirection
```
You could also call `ensureParquetLogRedirection` from some main class right? e.g. `class Executor`.

mallman replied:
I've tried to keep this code isolated to the parquet implementation.

mallman commented:
Are we planning to incorporate the Parquet 1.9 libraries into Spark 2.1? If so, then this PR should be unnecessary.

Hopefully.

mallman commented:
In fact, if no one else is working on the Parquet upgrade it probably makes more sense for me to contribute that then continue working on this PR. I'll check with the dev mailing list.

srowen commented:
Yeah there's a ticket for it to solve a different parquet bug. That might be the path of least resistance.

mallman commented:
I found two such tickets. How should we organize this in Jira?

srowen commented:
Ah right. Let me redirect 18140 to 13127 because all of these end up being resolved by "upgrade to 1.9"

rxin commented:
@mallman are you working on Parquet 1.9 upgrade? Would be great to get that in sooner than later because in the past Parquet upgrades tend to bring a lot of issues, so it'd be good to go through more QA testing.

mallman commented:
Yes, I am working on it. I'm planning to have a PR in no later than EOW.

mallman commented:
@rxin One of our weekly Spark jobs is choking, and fixing it could take the rest of the week. I suggest someone else take the lead on the Parquet 1.9 upgrade if they can devote their time to it.

cc @rdblue

mallman commented:
Hi Guys,

Unfortunately debugging our Spark job sucked up all my Spark time last week, and I still have more to do on it this week. Because of that, it doesn't look like I'll have time to work on the Parquet 1.9 upgrade after all.

FWIW, I did some work on it, got everything to compile and all the unit tests pass. However, I simply have to change my focus this week and can't devote time to another PR.

I'll rebase this patch and hope to get it in in case a 1.9 upgrade doesn't happen for Spark 2.1.

SparkQA commented:
**[Test build #68289 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68289/consoleFull)** for PR 15538 at commit [`34e3997`](https://github.com/apache/spark/commit/34e3997ed24ab37489620b0dda84a863ae3c2196).

mallman commented:
> If this is necessary, then isn't it simpler to leave the log configuration call where it was, so that it doesn't depend on the constructor? that wasn't the actual problem was it, just the logging? if the problem is that it fires repeatedly, then a flag could be used to record that it has executed?

Sorry @srowen, I don't understand what you're asking here. However, I can say that calling this method repeatedly is not the problem. The problem is getting it to be called before Parquet starts logging.

mallman commented:
I've pushed a rebase. I re-tested this PR using the methodology I describe in the description for both local and remote executors.

rxin commented:
@mallman are you also working on Parquet upgrade? If we don't get it in in the next day or two, we shouldn't merge that into branch-2.1 anymore since it is way past feature freeze.

srowen commented:
Yeah I see, but this is getting to be quite hacky just to turn off log messages, trying to make sure static init is triggered from a custom deserializer. It feels like this should be punted until Parquet 1.9 takes care of it.

mallman commented:
@rxin I'm not working on the Parquet upgrade this week. I think we'll have to punt on it.

SparkQA commented:
**[Test build #68289 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68289/consoleFull)** for PR 15538 at commit [`34e3997`](https://github.com/apache/spark/commit/34e3997ed24ab37489620b0dda84a863ae3c2196).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68289/
Test PASSed.

mallman commented:
> Yeah I see, but this is getting to be quite hacky **just to turn off log messages**

This isn't just a few annoying log messages. This is an _avalanche_ of log messages, each of which contains a deep stack trace.

You can read https://issues.apache.org/jira/browse/SPARK-17993?focusedCommentId=15645322&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15645322 for another perspective.

I agree this patch is ugly. I don't like writing code like this. I just don't see a feasible alternative for Spark 2.1. It can be removed as part of a future Parquet 1.9 upgrade.

I abbreviated the console output when I created SPARK-17993. Running a `spark-sql` query with 10 results on a table written with Spark 1.5 results in 7,052 lines of console output. (The last 10 are the actual results.)

srowen commented:
OK, maybe. It is not just about being hacky but being likely to break. If this really calls for a static initializer, how about a Java helper class to do the work? How about piggy backing on the Logging init routines? I don't see this as something without alternatives that needs to be pushed through

mallman commented:
Sorry, I don't see how Java code will make this patch any better. We're not really missing a static initializer. We just need _some_ initializer to run early enough.

If we can tolerate putting parquet-specific code in datasource-agnostic code there are a number of places where we can put a call to the log redirection code. I've tried to avoid putting parquet specific code outside the parquet specific codebase.

srowen commented:
If `ParquetFileFormat` had a static init block somehow, we'd be done right? because the logging config in that static initializer would have to execute, once, during classloading and therefore before any usage in a constructor or serialization. But per above, you can't write such a block in Scala. So, make a dummy `ParquetLogHelper.java` class, and make `ParquetFileFormat` have a dummy field of its type or something. `ParquetLogHelper` can have a static init, and must load when `ParquetFileFormat` loads. I might miss a reason that doesn't work, but seems like the sort of thing to rule out before settling on a complicated solution. It would be a little simpler than the current code.

I don't think it's so wrong to add logic directly to Logging.scala as a fairly temporary workaround. It wouldn't entail an actual compile or runtime dependency on Parquet in core. At least for me that seems simpler than the heroic attempt to intercept serialization and constructor paths. I'd hope it's sufficient to add it to Logging, which should have to init very early, everywhere, once.

mallman commented:
I'll work on a revision and try to push something today.

Commit:
Move Parquet log redirection code to a serializable Java class
ParquetLogRedirector which performs the redirection as part of its class
initialization

mallman commented:
Pushed a new version which I think is cleaner than before. I tested all 8 scenarios manually.

SparkQA commented:
**[Test build #68424 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68424/consoleFull)** for PR 15538 at commit [`247ef91`](https://github.com/apache/spark/commit/247ef91dc444bcba0751e9a7d76938c2bea07097).

SparkQA commented:
**[Test build #68424 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68424/consoleFull)** for PR 15538 at commit [`247ef91`](https://github.com/apache/spark/commit/247ef91dc444bcba0751e9a7d76938c2bea07097).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `final class ParquetLogRedirector implements Serializable `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68424/
Test PASSed.

srowen commented:
I like this. It centralizes and yet elaborates the handling of the issue, does seem to be the right mechanism to use, and I presume, does work reliably.

mallman commented:
@ericl Does this LGTY?

rxin commented:
I don't think we are going to upgrade Parquet for branch-2.1, since it's way past that point.  Let's merge this for 2.1.

Merging in master/branch-2.1.
